### DISCUSSION TITLE
Send Data to LED while Playing

### DISCUSSION THREAD

**Discussion by bmorse526** [schreibfaul1/ESP32-audioI2S#691] [Q&A] [answered]
I have the MP3 Decoding working from a Micro SD and playing to an I2S amplifier and being triggered by a PIR sensor.  This part is working fine.  I have another function that I would like to implement - 
Flash LED at rate of sound with brightness being determined by volume of current sound being played.

I found
```
void audio_process_extern(int16_t* buff, uint16_t len, bool *continueI2S){
  for(int i = 0; i < len; i += 2){
        int16_t data = buff[i];
       // do something with data.... }
}
```

And when I put it in the ESP32 project I can get the buffer length and data, But then data stream stops going to the I2S Amplifier.

I would like to be able to get a number that indicates the volume and send to the PCM processor to drive an LED while the stream is playing.  The only other option I have come up with is a second ESP32 to monitor the sound and to drive the LED - This should be able to be done in one set of hardware.  I appreciate any help or advice anyone could provide.

**Comment by schreibfaul1** [comment, marked as answer]
With audio_process_extern() you get every single sample in int16_t format.
````c++ 
uint16_t vum = audio.getVUlevel();
uint8_t left = vum >> 8; 
uint8_t right = vum & 0x00FF;
```` 
this gives you an average value for both channels, the values are between 0...127

**Comment by bmorse526** [comment]
That would be ideal.  But when I try to use it I get no member named getVUlevel exists.  All I can find is a getVolumn.  Is that the same call?

**Reply by schreibfaul1** [reply]
strange, the function must have been around for over a year
![image](https://github.com/schreibfaul1/ESP32-audioI2S/assets/26044260/b1b3bce2-449a-4527-b880-0bb8bf333c8d)
and is used in the MiniWebRadio project as a level indicator.

**Comment by bmorse526** [comment]
I think I found the issue.  The GIThub source code shows it.  I am using VS code for the ESP32 and the library it shows is 2.0.7 - GitHUB show 3.0.8P
Trying to figure out how to force it in VS Code I am using in Platform.ini
lib_deps =
  # RECOMMENDED
  # Accept new functionality in a backwards compatible manner and patches
  #  https://github.com/schreibfaul1/ESP32-audioI2S
  #esphome/ESP32-audioI2S @ 2.0.7
  esphome/ESP32-audioI2S @ 3.0.8P

I get SemanticVersionError: Invalid simple block '3.0.8p'

**Comment by bmorse526** [comment]
I tried using 
https://github.com/schreibfaul1/ESP32-audioI2S.git
Which is supposed to pull the latest from github.  But still get the SemanticVersionError: Invalid simple block '3.0.8p' so it will not compile and upload

**Comment by bmorse526** [comment]
Thank you for your help.  I got the Latest Library and cleared the History so now it is compiling and working.  

What is the levels that would be expected?  So far looks like 0 to over 32000.  Is there a built in max?

**Reply by schreibfaul1** [reply]
Yes, a uint16_t value is returned. This contains both channels "stacked on top of each other",
hence:

- x >> 8;
- x & 0x00FF;
